lib/libzfs/libzfs_pool.c

```diff
@@ -3854,6 +3854,10 @@ zpool_history_unpack(char *buf, uint64_t bytes_read, uint64_t *leftover,
 		if (ISP2(*numrecords + 1)) {
 			*records = realloc(*records,
 			    *numrecords * 2 * sizeof (nvlist_t *));
+			if (*records == NULL) {
```

Comment:
I'd like to rescind my previous comment.  While I don't care for the design of this at all, this code does seem to have been originally written to allow partial success on error.  Let's try and avoid changing those semantics and make a minimal fix.  If that means we need to add the tmp variable as in your original patch that's OK.